File: src/game/input/Action.java
package game.input;

/**
 * Created by thewa on 5/11/2016.
 */
public class Action {
    private OnClickListener onClickListener;
    private OnHoldListener onHoldListener;

    private boolean holding;
    private boolean pressed;
    private boolean clicked;

    private String name;

    public Action(String name) {
        this.name = name;
    }

    public Action() {

    }

    public void handleInput() {
        if (isHolding() && onHoldListener != null)
            onHoldListener.holding();
        if (isClicked() && onClickListener != null)
            onClickListener.clicked();
    }

    public void press() {
        if (!holding) {
            pressed = true;
            holding = true;
        }
        else pressed = false;
    }

    public void release() {
        if (holding || pressed)
            clicked = true;

        holding = false;
        pressed = false;
    }

    public boolean isHolding() {
        return holding;
    }

    public boolean isPressed() {
        if (pressed){
            pressed = false;
            return true;
        }
        return false;
    }

    public boolean isClicked() {
        if (clicked) {
            clicked = false;
            return true;
        }

        return false;
    }

    public String getName() {
        return name;
    }

    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public void setOnHoldListener(OnHoldListener onHoldListener) {
        this.onHoldListener = onHoldListener;
    }

    public interface OnClickListener {
        void clicked();
    }

    public interface OnHoldListener {
        void holding();
    }
}

File: src/game/input/Action.kt
package game.input

/**
 * Created by thewa on 5/11/2016.
 */
class Action {
    private var onClickListener: (() -> Unit)? = null
    private var onHoldListener: (() -> Unit)? = null

    var isHolding: Boolean = false
        private set
    private var pressed: Boolean = false
    private var clicked: Boolean = false

    val name: String

    constructor(name: String) {
        this.name = name
    }

    constructor() {
        name = ""
    }

    fun handleInput() {
        if (isHolding)
            onHoldListener?.invoke()
        if (isClicked)
            onClickListener?.invoke()
    }

    fun press() {
        if (!isHolding) {
            pressed = true
            isHolding = true
        } else
            pressed = false
    }

    fun release() {
        if (isHolding || pressed)
            clicked = true

        isHolding = false
        pressed = false
    }

    val isPressed: Boolean
        get() {
            if (pressed) {
                pressed = false
                return true
            }
            return false
        }

    val isClicked: Boolean
        get() {
            if (clicked) {
                clicked = false
                return true
            }

            return false
        }

    fun setOnClickListener(onClickListener: () -> Unit) {
        this.onClickListener = onClickListener
    }

    fun setOnHoldListener(onHoldListener: () -> Unit) {
        this.onHoldListener = onHoldListener
    }
}
